state/modelsummaries.go

```diff
@@ -0,0 +1,435 @@
+			SLALevel:           string(doc.SLA.Level),
+			SLAOwner:           doc.SLA.Owner,
+			CloudTag:           names.NewCloudTag(doc.Cloud).String(),
+			CloudRegion:        doc.CloudRegion,
+			CloudCredentialTag: cloudCred,
+			/// Users:              make(map[string]UserAccessInfo),
+			/// Machines:           make(map[string]MachineModelInfo),
+		}
+		p.indexByUUID[doc.UUID] = i
+		p.modelUUIDs[i] = doc.UUID
+	}
+	return p
+}
+
+func (p *modelSummaryProcessor) fillInFromConfig() error {
+	// We use the raw settings because we are reading across model UUIDs
+	rawSettings, closer := p.st.database.GetRawCollection(settingsC)
+	defer closer()
+
+	remaining := set.NewStrings(p.modelUUIDs...)
+	settingIds := make([]string, len(p.modelUUIDs))
+	for i, uuid := range p.modelUUIDs {
+		settingIds[i] = uuid + ":" + modelGlobalKey
+	}
+	query := rawSettings.Find(bson.M{"_id": bson.M{"$in": settingIds}})
+	var doc settingsDoc
+	iter := query.Iter()
+	for iter.Next(&doc) {
+		idx, ok := p.indexByUUID[doc.ModelUUID]
+		if !ok {
+			// How could it return a doc that we don't have?
+			continue
+		}
+		remaining.Remove(doc.ModelUUID)
+
+		cfg, err := config.New(config.NoDefaults, doc.Settings)
+		if err != nil {
+			// err on one model should kill all the other ones?
+			return errors.Trace(err)
+		}
+		detail := &(p.summaries[idx])
+		detail.ProviderType = cfg.Type()
+		detail.DefaultSeries = config.PreferredSeries(cfg)
+		if agentVersion, exists := cfg.AgentVersion(); exists {
+			detail.AgentVersion = &agentVersion
+		}
+	}
+	if err := iter.Close(); err != nil {
+		return errors.Trace(err)
+	}
+	if !remaining.IsEmpty() {
+		// XXX: What error is appropriate? Do we need to care about models that its ok to be missing?
+		return errors.Errorf("could not find settings/config for models: %v", remaining.SortedValues())
+	}
+	return nil
+}
+
+// removeExistingUsers takes a set of usernames, and removes any of them that are known-valid users.
+// it leaves behind names that could not be validated
+func (p *modelSummaryProcessor) removeExistingUsers(names set.Strings) error {
+	// usersC is a global collection, so we can access it from any state
+	users, closer := p.st.db().GetCollection(usersC)
+	defer closer()
+
+	var doc struct {
+		Id string `bson:"_id"`
+	}
+	query := users.Find(bson.M{"_id": bson.M{"$in": names.Values()}}).Select(bson.M{"_id": 1})
+	query.Batch(100)
+	iter := query.Iter()
+	for iter.Next(&doc) {
+		names.Remove(doc.Id)
+	}
+	if err := iter.Close(); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (p *modelSummaryProcessor) fillInFromStatus() error {
+	// We use the raw statuses because otherwise it filters by model-uuid
+	rawStatus, closer := p.st.database.GetRawCollection(statusesC)
+	defer closer()
+	statusIds := make([]string, len(p.modelUUIDs))
+	for i, uuid := range p.modelUUIDs {
+		statusIds[i] = uuid + ":" + modelGlobalKey
+	}
+	// TODO: Track remaining and error if we're missing any
```

Comment:
If we plan to leave TODOs in this code, could we please follow the format TODO (name)...
Although there is 'git blame', the code could be shuffled and then it can become harder to track the person with the context :)